### PR TITLE
: Tokens

### PR DESCRIPTION
**Pull request by andreytrofimov**

### CODE REVIEW

File: tests/react-ioc.test.tsx [hunk 1]
```diff
@@ -31,13 +38,15 @@ describe('[IOC tests]', () => {
...
-            <Provider singletons={[SampleStore]}>
+            <Provider singletons={[SampleStore, { token: ANOTHER_STORE_TOKEN, provider: AnotherStore }]}>
                 <SampleConsumer setContainer={setContainer} />
             </Provider>
         ));
 
         expect(contextContainer.get(SampleStore)).toBeDefined();
         expect(contextContainer.get(SampleStore).singletonVariable).toBe(2);
+        expect(contextContainer.get(ANOTHER_STORE_TOKEN)).toBeDefined();
+        expect(contextContainer.get<SampleStore>(ANOTHER_STORE_TOKEN).singletonVariable).toBe('3');
```
Comment (kanoshin) [inline tests/react-ioc.test.tsx]:
```suggestion
        expect(contextContainer.get<AnotherStore>(ANOTHER_STORE_TOKEN).singletonVariable).toBe('3');
```

File: src/react-ioc.tsx [hunk 2]
```diff
@@ -6,20 +6,44 @@ import 'reflect-metadata';
...
+}
+
+type ProviderConfigEntry<T> =
+    | interfaces.Newable<T>
+    | ClassProviderConfigEntry<T>
+    | ValueProviderConfigEntry<T>
+    ;
+
+type ClassProviderConfigEntry<T> = {
+    provide: SymbolToken<T> | interfaces.Newable<T>,
```
Comment (kanoshin) [inline src/react-ioc.tsx]: What is `| interfaces.Newable<T>` for?

File: src/react-ioc.tsx [hunk 3]
```diff
@@ -76,30 +100,45 @@ class ProviderComponent
...
         }
         if (this.props.instances) {
-            this.props.instances.forEach((store) => {
-                this.container.bind(store).to(store);
-            });
+            this.props.instances.forEach(this.bind);
         }
     }
 
+    bind = (pce: ProviderConfigEntry<any>): interfaces.BindingInSyntax<any> => {
```
Comment (kanoshin) [inline src/react-ioc.tsx]: `pce` is a very cryptic name.

File: src/react-ioc.tsx [hunk 4]
```diff
@@ -76,30 +100,45 @@ class ProviderComponent
...
+        if (isClassProviderConfigEntry(pce)) {
+            return this.container.bind(pce.provide).to(pce.useClass);
+        }
+
+        if (isValueProviderConfigEntry(pce)) {
+            // using toDynamicValue to return BindingInSyntax
```
Comment (kanoshin) [inline src/react-ioc.tsx]: Could you clarify a bit more? What is BindingInSyntax and why `toConstantValue` can't be used?

Comment (andreytrofimov) [reply src/react-ioc.tsx]: BindingInSyntax is interface which has inSingletonScope()
https://github.com/servicetitan/react-ioc/blob/1c250f863cb29cb0535b139387d760c12af43d25/src/react-ioc.tsx#L103-L105

File: src/react-ioc.tsx (hunk 3 above)
Comment (andreytrofimov) [reply src/react-ioc.tsx]: Renamed to `entry`

File: src/react-ioc.tsx (hunk 2 above)
Comment (andreytrofimov) [reply src/react-ioc.tsx]: Sort of premature implementation.
```tsx
<Provider
    singletons={[LogService]}
>
    <Provider 
        singletons={[{ provide: LogService, useClass: BetterLogService ]}
    >
        <LogServiceConsumer />
    </Provider>
</Provider
```

Comment (andreytrofimov) [reply src/react-ioc.tsx]: Not tested yet . Can remove for now

File: src/react-ioc.tsx [hunk 5]
```diff
@@ -6,20 +6,22 @@ import 'reflect-metadata';
...
-};
 
 // Omit taken from https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html
 export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
 
+type ServiceProvider<T = any> = interfaces.Newable<T> | TokenProvider<T>;
+
+export type Token<P> = interfaces.Newable<P> | symbol | string;
+
+type TokenProvider<T = any> = { token: Token<T>, provider: T };
```
[the remark this answers is not in the document]
Comment (andreytrofimov) [reply src/react-ioc.tsx]: @xarety 
After a discussion with Kirill, we changed this type. Using as an example https://angular.io/guide/dependency-injection-providers

File: src/react-ioc.tsx [hunk 6]
```diff
@@ -6,20 +6,44 @@ import 'reflect-metadata';
 export { interfaces };
 
 export { rootContainer };
-export type Newable = {
-    new (...args: any[]): any;
+
+// Omit taken from https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-8.html
+export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;
```
Comment (andreytrofimov) [inline src/react-ioc.tsx]: @kanoshin 
Do we still need this?

File: src/react-ioc.tsx (hunk 4 above)
Comment (kanoshin) [reply src/react-ioc.tsx]: Got it. This made me think that we have an excessive API, there is no difference between providing value as `instances` or `singletons`:
```
// both approaches of providing are doing the same
<Provider singletons={[{ provide: ANOTHER_STORE_TOKEN, useValue: value }]} instances={[{ provide: ANOTHER_STORE_TOKEN, useValue: value }]} />
```

Comment (kanoshin) [reply src/react-ioc.tsx]: We can leave `toDynamicValue` for now, but it seems we need to rethink our providing API.

File: src/react-ioc.tsx (hunk 2 above)
Comment (kanoshin) [reply src/react-ioc.tsx]: I think what confused me more is that `ValueProviderConfigEntry` only allows symbols for `provide` part, I expected it to be the same in `ClassProviderConfigEntry`.

File: src/react-ioc.tsx (hunk 6 above)
Comment (kanoshin) [reply src/react-ioc.tsx]: Or if it is no longer used.

Comment (andreytrofimov) [reply src/react-ioc.tsx]: It's no longer used, but it's exported

Comment (kanoshin) [reply src/react-ioc.tsx]: Let's remove it.

File: src/react-ioc.tsx (hunk 2 above)
Comment (xarety) [inline src/react-ioc.tsx]:
```suggestion
    provide: Token<T>,
```

Comment (xarety) [reply src/react-ioc.tsx]: What's about make `ValueProviderConfigEntry` same as `ClassProviderConfigEntry` in that case e.g we can provide configuration through an abstract class with a scheme of this configuration?

Comment (andreytrofimov) [reply src/react-ioc.tsx]: `useValue ` reduces syntax noise for those cases, and only costs about 15 lines. So I am for `useValue`

Comment (xarety) [reply src/react-ioc.tsx]: Ah, I incorrectly wrote my idea, I suggesting next type:
```
type ValueProviderConfigEntry<T> = {
    provide: Token<T>,
    useValue: T
};
```
in that case, we could provide values through abstract classes. I didn't suggest rid of `useValue`. 🙂

Comment (andreytrofimov) [reply src/react-ioc.tsx]: Do you mean to use an abstract class as a token for constant values? Could you provide an example?

Comment (xarety) [reply src/react-ioc.tsx]: Ah, ffs, an abstract class isn't `Newable`, I wanted something like this:
```
abstract class AppConfiguration {
    title!: string;
}

const App: React.FC = provide({
    singletons: [
        { provide: AppConfiguration, useValue: { title: 'ServiceTitan' } }
    ]
})(() => {
    const [configuration] = useDependencies(AppConfiguration);

    return <h1>`Welcome to ${configuration.title}`</h1>;
});
```

Comment (andreytrofimov) [reply src/react-ioc.tsx]: Got it. It looks clearer than
```ts
interface IAppConfiguration {
    title!: string;
}
const AppConfiguration = symbolToken<IAppConfiguration>('AppConfiguration');
```
It limits the possible values to objects, but I like it.

Comment (andreytrofimov) [reply src/react-ioc.tsx]: @kanoshin ?

Comment (kanoshin) [reply src/react-ioc.tsx]: Let's support both approaches. Abstract classes might be useful in cases when partial implementation is provided by our packages e.g. `LogService`. Not sure about using abstract class as a token for plain objects, feels somewhat unconventional.

**Comment by andreytrofimov** [comment]
Updated [README.md](https://github.com/servicetitan/react-ioc/blob/tokens/README.md)

**Pull request merged**